src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -194,11 +194,13 @@ public void OnCompleted(Action continuation)
             }
             else if (awaitableState == _awaitableIsCompleted)
             {
-                ThreadPool.QueueUserWorkItem((o) => ((Action)o)(), continuation);
+                // Read already completed, continue on calling thread
+                continuation();
```

Comment:
From a quick glance, this is suspectible.  An awaiter's OnCompleted should always queue the invocation of the delegate.